### ISSUE TITLE
Signature help fails on method defined below current line

### ISSUE THREAD

**Issue by fannheyward** [microsoft/pyright#1251]
**Describe the bug**

First reported at https://github.com/fannheyward/coc-pyright/issues/177. `self.method2(` in method1 it will show autocomplete but signature help fails.

**To Reproduce**

```python
class Test:
    def method1(self, param1):
        self.method2( // no signature help shows
        pass

    def method2(self, param1):
        self.method1( // signature help shows
        pass
```

**VS Code extension or command-line**

VSCode Pyright Version: 1.1.93

**Comment by erictraut** [comment]
I'm not able to repro the problem. Here's what I see when I type the open parenthesis in both locations:

![Screen Shot 2020-12-07 at 6 12 32 PM](https://user-images.githubusercontent.com/7040122/101429476-e5c51080-38b7-11eb-8aef-caac81e85097.png)

![Screen Shot 2020-12-07 at 6 12 24 PM](https://user-images.githubusercontent.com/7040122/101429480-e78ed400-38b7-11eb-987a-58236e063e51.png)

**Comment by erictraut** [comment]
Note that I'm using VS Code. If you're using a different editor, then it's probably a problem with that client, not with Pyright.

**Comment by fannheyward** [comment]
You are using VSCode auto closing brackets, you should disable it `editor.autoClosingBrackets`.

<img width="503" alt="截屏2020-12-08 10 25 41" src="https://user-images.githubusercontent.com/345274/101430329-bb1e9a80-393f-11eb-9ea7-82c5f2435ab3.png">

With this disabled, the signature help will not be triggered after input `(`.

**Comment by fannheyward** [comment]
<img width="308" alt="截屏2020-12-08 10 26 58" src="https://user-images.githubusercontent.com/345274/101430412-e86b4880-393f-11eb-9429-e5aea4d60b37.png">
<img width="434" alt="截屏2020-12-08 10 27 25" src="https://user-images.githubusercontent.com/345274/101430449-f8832800-393f-11eb-96a5-f1c9e704e23d.png">

**Comment by erictraut** [comment]
Thanks, I'm able to repro the problem now. The issue is that without the closing paren, the parser is unable to produce a parse tree that the signature help can use. I'll see if I can make the parse recovery logic more robust in this particular case.

**Comment by erictraut** [comment]
Parse recovery for an unterminated parenthesis in the Python language is very difficult because the tokenization rules change while you're within parentheses. Newlines, indent and dedent tokens are swallowed and not reported to the parser. 

In your example, the unterminated parenthesis causes the parser to consume the "pass" identifier (which is normally a keyword but is a legitimate variable name in this context). It then sees the "def" identifier which is not allowed by the grammar, so it reports a syntax error. But there's no way to recover the parse in a satisfactory way at this point because the newline and dedent that would normally be required to complete the `method1` function scope are swallowed by the tokenizer because it's in "parentheses mode". The `method2` function declaration is therefore invalid, which is why you don't see it appear as a suggested signature.

I don't see any good workaround for this problem. I suspect that no one has reported it previously because it's very unusual for Python developers to disable the "Auto Closing Brackets" feature. Is there a reason why you have this option disabled?

**Comment by fannheyward** [comment]
> Is there a reason why you have this option disabled?

No, I use this enabled, and I use coc-pairs in vim to do the same thing. I disabled it just to reproduce this issue.

cc @SvanT, please take a look at this issue.

**Comment by erictraut** [comment]
I think we'll need to mark this as "as designed" and close it. I don't see a good way to fix it given the limitations of the Python grammar and tokenization rules. The workaround is to recommend that everyone use "Auto Closing Brackets" or the comparable feature within their editor.

**Comment by SvanT** [comment]
Thank you guys for pinning this down so quickly! I'll use auto bracket close then 🙂
